Mathlib/RingTheory/Regular/IsSMulRegular.lean

```diff
@@ -0,0 +1,203 @@
+/-
+Copyright (c) 2024 Brendan Murphy. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Brendan Murphy
+-/
+import Mathlib.Algebra.Module.Torsion
+import Mathlib.RingTheory.Flat.Basic
+import Mathlib.RingTheory.Ideal.AssociatedPrime
+import Mathlib.RingTheory.Regular.ModSMulBy
+
+/-!
+# Lemmas about the `IsSmulRegular` Predicate
+
+For modules over a ring the proposition `IsSmulRegular r M` is equivalent to
+`r` being a *non zero-divisor*, i.e. `r • x = 0` only if `x = 0` for `x ∈ M`.
+This specific result is `isSMulRegular_iff_smul_eq_zero_imp_eq_zero`.
+Lots of results starting from this, especially ones about quotients (which
+don't make sense without some algebraic assumptions), are in this file.
+We don't pollute the `Mathlib.Algebra.Regular.SMul` file with these because
+it's supposed to import a minimal amount of the algebraic hierarchy.
+
+## Tags
+
+module, regular element, commutative algebra
+-/
+
+section Congr
+
+variable {R S M N} [Semiring R] [Semiring S] {σ : R →+* S} {σ' : S →+* R}
+    [RingHomInvPair σ σ'] [RingHomInvPair σ' σ] [AddCommMonoid M] [Module R M]
+
+lemma LinearEquiv.isSMulRegular_congr' [AddCommMonoid N] [Module S N]
+    (e : M ≃ₛₗ[σ] N) (r : R) : IsSMulRegular M r ↔ IsSMulRegular N (σ r) :=
+  e.toEquiv.isSMulRegular_congr (e.map_smul' r)
+
+lemma LinearEquiv.isSMulRegular_congr [AddCommMonoid N] [Module R N]
+    (e : M ≃ₗ[R] N) (r : R) : IsSMulRegular M r ↔ IsSMulRegular N r :=
+  e.isSMulRegular_congr' r
+
+end Congr
+
+namespace IsSMulRegular
+
+open Submodule
+open scoped Pointwise TensorProduct
+
+variable {R M M' M'' : Type*}
+
+lemma submodule [Semiring R] [AddCommMonoid M] [Module R M]
+    (N : Submodule R M) (r : R) (h : IsSMulRegular M r) : IsSMulRegular N r :=
+  h.of_injective N.subtype N.injective_subtype
+
+section Ring
+
+variable (M) [Ring R] [AddCommGroup M] [Module R M]
+    [AddCommGroup M'] [Module R M'] [AddCommGroup M''] [Module R M'']
+    (N : Submodule R M) (r : R)
+
+lemma isSMulRegular_iff_smul_eq_zero_imp_eq_zero :
+    IsSMulRegular M r ↔ ∀ x : M, r • x = 0 → x = 0 :=
+  Iff.trans (Module.toAddMonoidEnd R M r).ker_eq_bot_iff.symm
+    <| AddSubgroup.eq_bot_iff_forall _
```

Comment:
I think you can use `nonZeroSMulDivisors` here?

Reply:
I could, but I don't see the benefit of doing so. There's a benefit to stating things explicitly, changing it breaks at least one existing proof, and atm there's nowhere it's useful to have the set `nonZeroSMulDivisors`

Reply:
> There's a benefit to stating things explicitly

I don't see that here though.

> there's nowhere it's useful to have the set `nonZeroSMulDivisors`

I think this is because `nonZeroSMulDivisors` is lacking API, but arguably some of the lemmas below (for example `isSMulRegular_of_range_eq_ker`) has nothing to do with the ring structure of `R` and is actually about `nonZeroSMulDivisors`.

Whether the existing API can help you is only half the story though, the other half is whether your code could have improved the existing API.

I'll be satisfied if you add a separated lemma `IsSMulRegular M r ↔ r ∈ nonZeroSMulDivisors R M` though.
At least people searching for API around `nonZeroSMulDivisors` can find this page.